GUI: Provide mechanism to disable gui overlay when empty

popcornmix opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
GUI: Provide mechanism to disable gui overlay when empty

## Motivation and Context
1080p60 requires 480MB/s of sdram bandwidth to display the Kodi GUI.
When playing a video that doesn't use subtitles or have the OSD active this bandwidth
is just wasted fetching transparent pixels.
Benchmarks show about 10% better performance when software decoding HEVC on a Pi3
when this bandwidth is not being consumed.

## How Has This Been Tested?
This has been in recent Milhouse builds and improved HEVC performance has been observed.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

popcornmix committed:
rbp: Remove GUI layer during video playback when it is empty

FernetMenta commented on xbmc/Application.cpp:
```diff
@@ -2737,6 +2737,7 @@ void CApplication::FrameMove(bool processEvents, bool processGUI)
       {
         CRenderingStats stats;
         g_windowManager.Process(CTimeUtils::GetFrameTime(), stats);
+        g_Windowing.SetVisible(stats.visible > 0);
```
I would rather see pi polling this info from guilib than creating a new interface only a single platform will use.

xhaggi commented:
Should it be enough to introduce a method in our window manager which return if there are at least one  visible window/dialog instead of passing this stats object everywhere?

popcornmix commented:
@xhaggi all I need to know is if gui is completely transparent (i.e nothing is visible).
If you know an easier way than getting stats from Process calls then let me know how.

FernetMenta commented:
knowing how many controls are visible/no visiable/dirty could be an advantage in many cases. just recently I fixed this high CPU load bug because of a control that always signalled dirty. those kind of isses are hard to track down. this new state cold make it a lot easier.

xhaggi commented:
After a deeper look your approach, counting the controls in `Process`, is totally fine. And what @FernetMenta said makes totally sense. We should add the stats class as a member to our window manager and expose it via a public method. That way you can easily access the stats.

popcornmix commented:
@FernetMenta @xhaggi I've changed the scheme to collect stats in CGUIWindowManager and poll this from MMAL renderer. Is this preferable?

FernetMenta commented on xbmc/windowing/WinSystem.h:
```diff
@@ -85,6 +85,9 @@ class CWinSystemBase
   virtual bool Show(bool raise = true) { return false; }
+  // Indicates if GUI layer is currently visible. When playing video it may be completely empty
+  // and can be disabled for performance reasons
+  virtual void SetVisible(bool visible) {}
```
I still don't like this method. it is very unlikely that this will be ever used by other ports. pi windowing system can poll WindowManager

MrMC replied:
Other Embedded apps can use it. Desktop apps have the ponies and just don't care.

FernetMenta replied:
"can" but most likely "won't". you need to look at the codebase from an overall perspective. platform specific tweaks made the codebase rather brittle in the past.
the problem here can be solved easily without this additional interface method. if the 3rd or 4th port in the future will make use of this feature, we can bring this up again.

FernetMenta commented on xbmc/cores/VideoPlayer/VideoRenderers/HwDecRender/MMALRenderer.cpp:
```diff
@@ -821,6 +823,9 @@ void CMMALRenderer::RenderUpdate(bool clear, DWORD flags, DWORD alpha)
   int source = m_iYV12RenderBuffer;
   CMMALBuffer *omvb = nullptr;
 
+  CRenderingStats stats = g_windowManager.GetRenderingStats();
```
this thread will finally call Windowing::PresentRender which can be overwritten by the Pi port. then there is no need to have g_Windowing.SetVisible

xhaggi commented on xbmc/guilib/DirtyRegion.h:
```diff
@@ -35,3 +35,10 @@ class CDirtyRegion : public CRect
 typedef std::vector<CDirtyRegion> CDirtyRegionList;
+
+class CRenderingStats
+{
```
Any reason why this is part of the DirtyRegion header? Should we move it to its own header or move it to window manager header?

popcornmix replied:
GUIWindowManager.h isn't currently included by the GUI controls.

It is currently here as CDirtyRegionList is the other parameter to Process and the stats are comparable to the dirty region list (i.e. info on what will be rendered).
I can move it but adding a new include to each GUI control is somewhat painful so I'd like to sure it of the right place before doing it.

xhaggi replied:
Okay. What do you think about if we create a new container class where we put all the stuff (dirty time, dirty region list, render stats) in, something​ like `ProcessInfo` and use it as parameter in `Process`. This way we don't need to alter the method signature of `Process` again and again if we need to add something else, in the future Then we could rename the header file too.

popcornmix replied:
@xhaggi I assume you are not referring to the existing ``ProcessInfo.h`` (used by VideoPlayer)? So we'll probably need a different name to avoid confusion.

FernetMenta replied:
good idea. I am fine if this is done in this PR or a subsequent one.

xhaggi replied:
@popcornmix sure don't know that there's already a class named `ProcessInfo`

Paxxi commented:
Why would this only apply to the pi? Wasting cpu cycles is bad on every platform so this should just be sort of the default case imho. Don't compute/render stuff that's not seen/visible.

FernetMenta commented:
@Paxxi you seem not to know what this PR is about. it turns off gui rendering during video playback if no gui controls are visible. this only affects bypass renderers that have a separate video surface for output of video. this is not default. default is that video is rendered to the same surface as gui

Paxxi commented:
So no other platform does the empty render/transfer?

fritsch commented:
Let's state it this way - the majority of kodi users is running bypass
renderers :-) - it's namely Android and the PI  (and IMX and AMLogic). It
would be great if such platforms could benefit from those improvements out
of the box.

2017-05-05 20:49 GMT+02:00 Pär Björklund <notifications@github.com>:

> So no other platform does the empty render/transfer?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/xbmc/xbmc/pull/12054#issuecomment-299546023>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABCfHZbCv-UAADChnaANLDQ7W5OADKyzks5r228ngaJpZM4NQyzz>
> .
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

popcornmix commented:
@FernetMenta @xhaggi I've updated PR with (some of) your suggestions.
I've rebased it onto https://github.com/xbmc/xbmc/pull/11986 as the rpi windowing code is better separated, so ignore the initial commits from there.

FernetMenta commented:
@popcornmix thanks this is much much better.

@fritsch you seem not to have understood the probem.

fritsch commented:
It seems you did not explain it properly then.  But in fact I don't understand why it does not affect android's surface rendering and also why not in the IMX case. Mind explaining?

FernetMenta commented:
Why don't you try to switch off Android's gui rendering layer. I doubt that this is even possible. Disabling the main rendering layer on an active GL/GLES context is a odd and very specific case. There is nothing in this context you get out of the box.

fritsch commented:
Thanks for the explanation on IRC. I mixed that up with a better version for "only render the gui 15 fps" if video is playing. Thanks much.

xhaggi commented on xbmc/windowing/rpi/RPIUtils.cpp:
```diff
@@ -68,6 +68,10 @@ CRPIUtils::CRPIUtils()
   m_height = 1280;
   m_width = 720;
+  m_screen_width   = 1280;
+  m_screen_height  = 720;
```
Superfluous spaces, really minor.

xhaggi commented:
@popcornmix thank you, much better now

ksooo commented on xbmc/guilib/DirtyRegion.h:
```diff
@@ -35,3 +35,14 @@ class CDirtyRegion : public CRect
+  unsigned int currentTime;
+  CDirtyRegionList dirtyregions;
+  int visible;
+  int total;
```
A class with public members? Wait - wasn't this all about encapsulating data? ;-) I would either use a struct, if encapsulation is not wanted here, or a class to give CGUIRenderInfo control of the data.

xhaggi replied:
It would be great to have methods to access the member and make them private, like `incrementVisible()`, `addDirtyRegion()` and the necessary getter, but no strong opinion here.

ksooo replied:
> It would be great to have methods to access the member and make them private

 Well, not exactly "great", but OOP. ;-)

FernetMenta replied:
> It would be great to have methods to access the member and make them private

that would just inrease cpu load in this case. assume you have 100 controls on screen and you have to call 3-4 methods on every control. evaluating dirty regions should not get more expensive that rendering the whole area

Montellese replied:
simple methods like getters and increments can be inlined by the compiler so you shouldn't loose anything there

FernetMenta replied:
please refrain from bikeshedding here. the PR is ok in current state. if anyone thinks that getter/setter would improve things in that area, please submit a PR along with performance measures.

ksooo replied:
Using classes with public members  is never good design. Make it a struct if you don't need/want data encapsulation. This is c++ basics and no bike shedding, imo.

FernetMenta replied:
@popcornmix could you make it a struct?

xhaggi replied:
If we have a struct now should we remove the `C` from the name. IMO leave it a class with public members for now or add the missing inline methods if you have the time. It'll be easier to adjust the class in a second round as changing from struct to class over again. Just my two cents.

ksooo commented on xbmc/guilib/GUIBaseContainer.cpp:
```diff
@@ -68,9 +68,9 @@ CGUIBaseContainer::~CGUIBaseContainer(void)
 }
 
-void CGUIBaseContainer::DoProcess(unsigned int currentTime, CDirtyRegionList &dirtyregions)
+void CGUIBaseContainer::DoProcess(CGUIRenderInfo &renderInfo)
```
Just a question: why are all the new params not <code>const CGUIRenderInfo &</code>? Is render info supposed to be altered by that methods? If not, I would go for const wehenever possible.

FernetMenta replied:
> Is render info supposed to be altered by that methods

sure, this is the only reason for this method

ksooo replied:
Ah, okay, thanks.

popcornmix commented:
I've rebased and returned to class with private members. I'm happy the compiler will inline them.
Is this preferable?

xhaggi commented:
Greet, thank you. No objections from my side.

FernetMenta commented:
imo setter/getters are nonsens here. the struct was much better

ksooo commented:
> I've rebased and returned to class with private members. I'm happy the compiler will inline them.
Is this preferable?

+1

ksooo commented on xbmc/guilib/DirtyRegion.h:
```diff
@@ -35,3 +35,23 @@ class CDirtyRegion : public CRect
+  void IncTotal() { m_total++; }
+  int GetVisible() { return m_visible; }
+  int GetTotal() { return m_total; }
+  unsigned int GetTime() { return m_currentTime; }
```
The getters should be const. For example <code>unsigned int GetTime() const</code>.

ksooo commented on xbmc/guilib/DirtyRegion.h:
```diff
@@ -35,3 +35,23 @@ class CDirtyRegion : public CRect
+  unsigned int GetTime() { return m_currentTime; }
+  void AddRegion(const CDirtyRegion &region) { m_dirtyregions.push_back(region); }
+  CDirtyRegionList &GetRegions() { return m_dirtyregions; }
+  size_t GetRegionSize() { return m_dirtyregions.size(); }
```
All the getters should be const. For example unsigned int GetTime() const.

peak3d commented:
pls.wait with this one. I'm currently trying to supress rendering / processing of the GUI completely if no gui update is necessary

peak3d commented on xbmc/guilib/GUIControl.cpp:
```diff
@@ -124,21 +124,25 @@ void CGUIControl::DynamicResourceAlloc(bool bOnOff)
-  changed |= Animate(currentTime);
+  changed |= Animate(renderInfo.GetTime());
 
+  if (m_controlID != WINDOW_FULLSCREEN_VIDEO)
```
I dont like to have this special case in CGUIControl, its against OOP.
It should be solved using virtual member (IncVisibleCount()) in CGUIControl and should be overwritten in CGUIWindowFullScreen.

peak3d commented on xbmc/guilib/GUIControl.cpp:
```diff
@@ -124,21 +124,25 @@ void CGUIControl::DynamicResourceAlloc(bool bOnOff)
   if (IsVisible())
   {
+    if (m_controlID != WINDOW_FULLSCREEN_VIDEO && !g_windowManager.IsAddonWindow(m_controlID) && !g_windowManager.IsPythonWindow(m_controlID))
+      renderInfo.IncVisible();
```
visibility expressions from skin should be checked as well.
IsVisible() only means that its "general visible", the visibility rules can change from true -> false

peak3d commented:
I'm introducing an alternative implementation here:

https://github.com/xbmc/xbmc/pull/12213/commits/11ecf27a85cf573d14f369777326718eafc5d1ce#diff-02cbf300ebd15bda2591e96d935b0f87R1607

MartijnKaijser commented:
@popcornmix this needs rebasing

popcornmix commented:
This PR is a lot simpler after https://github.com/xbmc/xbmc/pull/12213 and only affects Pi specific code.

jenkins build this please

JimmyS83 commented:
> 
> 
> Thanks for the explanation on IRC. I mixed that up with a better version for "only render the gui 15 fps" if video is playing. Thanks much.

I miss "only render the gui 1 fps" if black screensaver is active :-)

MartijnKaijser commented:
Please ask questions on the forum if they are not code related
